polygeist/tools/cgeist/Lib/ValueCategory.cc

```diff
@@ -281,12 +281,106 @@ template <typename OpTy> inline void warnUnconstrainedCast() {
 
 ValueCategory ValueCategory::FPTrunc(OpBuilder &Builder,
                                      Type PromotionType) const {
+  assert(val.getType().isa<FloatType>() &&
+         "Expecting floating point source type");
+  assert(PromotionType.isa<FloatType>() &&
+         "Expecting floating point promotion type");
+  assert(val.getType().getIntOrFloatBitWidth() >=
+             PromotionType.getIntOrFloatBitWidth() &&
+         "Source type must be wider than promotion type");
+
   warnUnconstrainedCast<arith::TruncFOp>();
   return Cast<arith::TruncFOp>(Builder, PromotionType);
 }
 
 ValueCategory ValueCategory::FPExt(OpBuilder &Builder,
                                    Type PromotionType) const {
+  assert(val.getType().isa<FloatType>() &&
+         "Expecting floating point source type");
+  assert(PromotionType.isa<FloatType>() &&
+         "Expecting floating point promotion type");
+  assert(val.getType().getIntOrFloatBitWidth() <=
+             PromotionType.getIntOrFloatBitWidth() &&
+         "Source type must be narrower than promotion type");
+
   warnUnconstrainedCast<arith::ExtFOp>();
   return Cast<arith::ExtFOp>(Builder, PromotionType);
 }
+
+ValueCategory ValueCategory::SIToFP(OpBuilder &Builder,
+                                    Type PromotionType) const {
+  assert(val.getType().isa<IntegerType>() && "Expecting int source type");
+  assert(PromotionType.isa<FloatType>() &&
+         "Expecting floating point promotion type");
+
+  warnUnconstrainedCast<arith::SIToFPOp>();
+  return Cast<arith::SIToFPOp>(Builder, PromotionType);
+}
+
+ValueCategory ValueCategory::UIToFP(OpBuilder &Builder,
+                                    Type PromotionType) const {
+  assert(val.getType().isa<IntegerType>() && "Expecting int source type");
+  assert(PromotionType.isa<FloatType>() &&
+         "Expecting floating point promotion type");
+
+  warnUnconstrainedCast<arith::UIToFPOp>();
+  return Cast<arith::UIToFPOp>(Builder, PromotionType);
+}
+
+ValueCategory ValueCategory::FPToUI(OpBuilder &Builder,
+                                    Type PromotionType) const {
+  assert(val.getType().isa<FloatType>() &&
+         "Expecting floating point source type");
+  assert(PromotionType.isa<IntegerType>() &&
+         "Expecting integer promotion type");
+
+  warnUnconstrainedCast<arith::FPToUIOp>();
+  return Cast<arith::FPToUIOp>(Builder, PromotionType);
+}
+
+ValueCategory ValueCategory::FPToSI(OpBuilder &Builder,
+                                    Type PromotionType) const {
+  assert(val.getType().isa<FloatType>() &&
+         "Expecting floating point source type");
+  assert(PromotionType.isa<IntegerType>() &&
+         "Expecting integer promotion type");
+
+  warnUnconstrainedCast<arith::FPToSIOp>();
+  return Cast<arith::FPToSIOp>(Builder, PromotionType);
+}
+
+ValueCategory ValueCategory::IntegerCast(OpBuilder &Builder, Type PromotionType,
```

Comment:
What's the benefit of having `ValueCategory::IntegerCast` instead of inlining the body to `ValueCategory::IntCast`?

Reply:
None, I was planning to fold casting of constants, but I decided that should be done at MLIR level. I'll inline. Thanks for pointing out.